src/vm/i386/jithelp.S

```diff
@@ -0,0 +1,691 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+.intel_syntax noprefix
+#include "unixasmmacros.inc"
+#include "asmconstants.h"
+
+// ***
+// JIT_WriteBarrier* - GC write barrier helper
+//
+// Purpose:
+//    Helper calls in order to assign an object to a field
+//    Enables book-keeping of the GC.
+//
+// Entry:
+//    EDX - address of ref-field (assigned to)
+//    the resp. other reg - RHS of assignment
+//
+// Exit:
+//
+// Uses:
+//    EDX is destroyed.
+//
+// Exceptions:
+//
+// *******************************************************************************
+
+//  The code here is tightly coupled with AdjustContextForWriteBarrier, if you change
+//  anything here, you might need to change AdjustContextForWriteBarrier as well
+.macro WriteBarrierHelper rg
+.align 4
+
+// The entry point is the fully 'safe' one in which we check if EDX (the REF
+// begin updated) is actually in the GC heap
+NESTED_ENTRY JIT_CheckedWriteBarrier\rg, _TEXT, NoHandler
+    // check in the REF being updated is in the GC heap
+    cmp     edx, g_lowest_address
```

Comment:
You'll need to use PREPARE_EXTERNAL_VAR macro for global variable access on Linux, like we do for AMD64.  Otherwise we would not get position independent code.